src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -10687,4 +10690,49 @@ template std::optional<EphemeralLockInZooKeeper> StorageReplicatedMergeTree::all
     const std::vector<String> & zookeeper_block_id_path,
     const String & zookeeper_path_prefix) const;
 
+Int32 StorageReplicatedMergeTree::tryFixMetadataVersionInZooKeeper()
+{
+    const Int32 metadata_version = getInMemoryMetadataPtr()->getMetadataVersion();
+    if (metadata_version != 0)
+    {
+        /// No need to fix anything
+        return metadata_version;
+    }
+
+    auto zookeeper = getZooKeeper();
+
+    Coordination::Stat stat;
+    zookeeper->get(fs::path(zookeeper_path) / "metadata", &stat);
+    if (stat.version == 0)
+    {
+        /// No need to fix anything
+        return metadata_version;
+    }
+
+    queue.pullLogsToQueue(zookeeper);
+    if (queue.getStatus().metadata_alters_in_queue != 0)
+    {
+        LOG_DEBUG(log, "No need to update metadata_version as there are ALTER_METADATA entries in the queue");
+        return metadata_version;
+    }
+
+    const Coordination::Requests ops = {
+        zkutil::makeSetRequest(fs::path(replica_path) / "metadata_version", std::to_string(stat.version), 0),
+        zkutil::makeCheckRequest(fs::path(zookeeper_path) / "metadata", stat.version),
+    };
+    Coordination::Responses ops_responses;
+    const auto code = current_zookeeper->tryMulti(ops, ops_responses);
+    if (code == Coordination::Error::ZOK)
+    {
+        LOG_DEBUG(log, "Successfully fixed metadata_version");
+        return stat.version;
+    }
+    if (code == Coordination::Error::ZBADVERSION)
+    {
+        LOG_DEBUG(log, "No need to update metadata_version because table metadata has been updated on a different replica");
```

Comment:
Let's check if executing an ALTER query on another replica actually fixes the issue

Reply:
I have refactored this method to retry if the ZooKeeper transaction failed with `ZBADVERSION`.

This can happen for two reasons:
* `metadata_version` became non-zero. Then on the second attempt we will return the new version.
* `metadata` changed during the first attempt. This implies that the replication log was also updated because these two operations are done in the same transaction. Then on the second attempt we will either get `metadata_version != 0` and return the new version or we will get `metadata_alters_in_queue != 0` and return 0.

Reply:
Okay, but will an ALTER query fix the issue? I mean, if we have an ALTER_METADATA entry in the replication queue, will the replica with uninitialized metadata_verion successfully execute it, and will it fix the metadata_version?

Reply:
Yes, it will set the entry's version: https://github.com/ClickHouse/ClickHouse/blob/fix-metadata-version-in-zookeeper/src/Storages/StorageReplicatedMergeTree.cpp#L5992